linux/system-config/create-snapshot

```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+
+# Usage: block-snapshot add|remove img-file cow-file
+#
+# This creates dm-snapshot device on given arguments
+
+SNAPSHOT_CHUNKSIZE=256  # same as in file.py
+
+set -euo pipefail
+shopt -s nullglob
+
+fatal () {
+  printf %s\\n "$@" >&2
+  exit 1
+}
+
+get_dev() {
+  dev=$1
+
+  if [ -L "$dev" ]; then
+    dev=$(readlink -f "$dev") || fatal "$dev link does not exist."
+  fi
+
+  if [ -f "$dev" ]; then
+    file=$dev
+
+    # assign new loop device
+    loopdev=$(losetup --find --nooverlap --show -- "$file" 2>/dev/null)
+    if [ $? -ne 0 ] || [[ ! "$loopdev" =~ ^/dev/loop[0-9]+$ ]]
+    then
+      fatal 'Failed to find an unused loop device'
+    fi
+
+    printf %s\\n "$loopdev"
+  else
+    test -e "$dev" || fatal "$dev does not exist."
+    test -b "$dev" || fatal "$dev is not a block device nor file."
+    printf %s\\n "$dev"
+  fi
+}
+
+file_to_devname () {
+  exec stat '--printf=-%D:%i' -- "$@"
+}
+
+create_dm_snapshot() {
+  local base_dev cow_dev base_sz dm_devname
+
+  dm_devname=$1
+  base_dev=$2
+  cow_dev=$3
+
+  if [ ! -e "/dev/mapper/$dm_devname" ]; then
+    # prepare new snapshot device
+    base_sz=$(blockdev --getsz "$base_dev")
+    3< "$base_dev" 4< "$cow_dev" dmsetup create "$dm_devname" \
+      --table "0 $base_sz snapshot /dev/fd/3 /dev/fd/4 P $SNAPSHOT_CHUNKSIZE" ||
```

Comment:
Using `/dev/fd` here doesn't help with anything (the device name is resolved by the shell just before calling `dmsetup`), but makes it harder to read. Please don't do that.